packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart

```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
   TimelineController() {
     timelineService = TimelineService(this);
-    fullTimeline = FullTimeline(this);
-    frameBasedTimeline = FrameBasedTimeline(this);
-    timelines = [frameBasedTimeline, fullTimeline];
+    processor = TimelineProcessor(this);
   }
 
   final cpuProfilerController = CpuProfilerController();
 
   /// Notifies that a timeline event was selected.
-  ValueListenable get selectedTimelineEventNotifier =>
-      _selectedTimelineEventNotifier;
+  ValueListenable get selectedTimelineEvent => _selectedTimelineEventNotifier;
   final _selectedTimelineEventNotifier = ValueNotifier<TimelineEvent>(null);
 
+  /// Notifies that a timeline frame has been selected.
+  ValueListenable get selectedFrame => _selectedFrameNotifier;
+  final _selectedFrameNotifier = ValueNotifier<TimelineFrame>(null);
+
+  /// Notifies when an empty timeline recording finishes
+  ValueListenable get emptyRecording => _emptyRecordingNotifier;
+  final _emptyRecordingNotifier = ValueNotifier<bool>(false);
+
+  /// Notifies that the timeline is currently being recorded.
+  ValueListenable get recording => _recordingNotifier;
+  final _recordingNotifier = ValueNotifier<bool>(false);
+
+  /// Notifies that the recorded timeline data is currently being processed.
+  ValueListenable get processing => _processingNotifier;
+  final _processingNotifier = ValueNotifier<bool>(false);
+
+  /// Stream controller that notifies the timeline has been processed.
+  Stream<bool> get onTimelineProcessed => _timelineProcessedController.stream;
+  final _timelineProcessedController = StreamController<bool>.broadcast();
+
   /// Stream controller that notifies that offline data was loaded into the
   /// timeline.
   ///
   /// Subscribers to this stream will be responsible for updating the UI for the
   /// new value of [timelineData].
-  final _loadOfflineDataController = StreamController<OfflineData>.broadcast();
+  final _loadOfflineDataController =
+      StreamController<OfflineTimelineData>.broadcast();
+
+  Stream<OfflineTimelineData> get onLoadOfflineData =>
+      _loadOfflineDataController.stream;
 
   /// Stream controller that notifies the timeline screen when a non-fatal error
   /// should be logged for the timeline.
   final _nonFatalErrorController = StreamController<String>.broadcast();
 
-  Stream<OfflineData> get onLoadOfflineData =>
-      _loadOfflineDataController.stream;
-
   Stream<String> get onNonFatalError => _nonFatalErrorController.stream;
 
   /// Stream controller that notifies the timeline has been cleared.
   final _clearController = StreamController<bool>.broadcast();
 
   Stream<bool> get onTimelineCleared => _clearController.stream;
 
-  TimelineBase get timeline =>
-      timelineModeNotifier.value == TimelineMode.frameBased
-          ? frameBasedTimeline
-          : fullTimeline;
-
-  FrameBasedTimeline frameBasedTimeline;
-
-  FullTimeline fullTimeline;
-
-  List<TimelineBase> timelines;
+  TimelineData data;
```

Comment:
we have both data and OfflineTimelineData.
Document expectations on whether one or both can be non-null.